Podcast Player: API endpoint improvements

marekhrabe opened the pull request:
This updates the helper class to format all data needed, not just tracks. And data from this class is reused in both API and server-side rendering.

#### Changes proposed in this Pull Request:
* More universal helper function to get all player data at once
* Error passing from helper to API
* New data loaded from the RSS:
  - podcast title
  - podcast homepage
  - podcast cover image
* API only accessible for logged-in users, with `edit_posts` capability

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
paYJgx-sH-p2

#### Testing instructions:
- smoke test block in the editor
  - should load episodes of podcast and those will be playable
  - inspect the API request in devtools and check it comes back with these keys (might be empty for some podcasts): `tracks`, `title`, `link` and `cover`
- test block on the fronted. should work as expected
- try accessing the endpoint directly, like http://localhost/wp-json/wpcom/v2/podcast-player?url=https%3A%2F%2Fdistributed.blog%2Fcategory%2Fpodcast%2Ffeed%2F
  - should fail with `rest_forbidden` because you are not authorized without token/cookie

#### Proposed changelog entry for your changes:
* none

marekhrabe committed:
update helper

marekhrabe committed:
use new helper function in api

marekhrabe committed:
use new data in edit.js

marekhrabe committed:
use new data on the frontend

marekhrabe committed:
check user in podcast api

jeherve commented on _inc/lib/class-jetpack-podcast-helper.php:
```diff
@@ -10,15 +10,58 @@
+			'title'  => $title,
+			'link'   => $link,
+			'cover'  => $cover,
+			'tracks' => $tracks,
```
We escape that data on the PHP side, but should we escape it here so we get escaped data in the player in the editor as well, when the data is queried via the endpoint?

Maybe this would also be a good place to add some default values when SimplePie returns `null` values for this info?

marekhrabe replied:
on it

marekhrabe replied:
I have updated the logic to have explicit null or empty strings based on the value and added text processing in order to account for both the editor (via API) and the server side render.

jeherve commented on _inc/lib/class-jetpack-podcast-helper.php:
```diff
@@ -10,15 +10,58 @@
 	 * @param int    $quantity Optional. The number of tracks to return.
+	 * @return array|WP_Error The player data or a error object.
+	 */
+	public static function get_player_data( $feed, $quantity = 10 ) {
```
Should we remove `$quantity` from here and from the block's PHP side now?

marekhrabe replied:
The server-side rendering still uses it to only inline data of tracks that will be used. 

We only cut the episode amount in the editor but the component on the frontend displays everything it gets.

marekhrabe replied:
I'm happy to leave quantity as-is, unless you object.

marekhrabe replied:
Update: I will be removing quantity in a follow up PR I have started.

marekhrabe committed:
add escaping and ensure some string are pure plaintext

marekhrabe commented:
@jeherve I took a look at escaping and default values and also addressed a separate issue where we have trouble with entities like `&amp;` and `&#039;` visually showing up in the player.

I have addressed this in a way:
- everything rendered in HTML markup (podcast-player.php) uses proper esc_ methods - no change there
- the helper generates clean content by:
  - calling esc_ functions where needed
  - cleaning other values from entities and HTML tags (to ensure they can be used as a plain text)

Ensuring our API returns plain text values is the best way to pass them into our React component. If we pass escaped content meant for HTML backend rendering (like `esc_html`), we will need to first decode it before placing in React using some library, or use `dangerouslySetInnerHTML`.

We have also initially used `wp_kses_post` for the episode description. This was an oversight - there is no HTML and we want just a plain text. It has been addressed in this PR too.

**Testing:**

Podcast https://feeds.feedburner.com/lerendezvousjeux is great to test with, as episode 134 contains an ampersand and episode 128 a single quote. Both were originally showed as entities.

Now they should both show up visually in the player in both the editor and frontend, not as their entity. Secondly, viewing the HTML of the generated frontend, you should be able to find the `<noscript>` tag the Podcast Player generates and ensure these two episode names include entities in their markup (as done by `esc_html`).

marekhrabe commented:
Rebased.

jeherve reviewed (approved):
This works nicely. It should be good to merge now!

marekhrabe commented:
r204805-wpcom
